### PR TITLE
Adding generic task for fetching a secure private repo

### PR DESCRIPTION
**Pull request by macdiesel**
@edx/pipeline-team PTAL.

### CODE REVIEW

File: edxpipelines/patterns/tasks.py [hunk 1]
```diff
@@ -307,6 +307,48 @@ def fetch_secure_configuration(job, secure_dir, runif="passed"):
     )
 
 
+def generate_fetch_secure_repository(job, repo_dir, runif="passed"):
+    """
+    Requires environment variables:
+        - PRIVATE_GITHUB_KEY            - the RSA key to use when cloning the github repo
+        - PRIVATE_REPOSITORY_URL        - URL used to clone the private repo
+        - PRIVATE_REPOSITORY_VERSION    - The hash or branch name to checkout
```
Comment (doctoryes) [inline edxpipelines/patterns/tasks.py]: This method doesn't work if you need to check out more than one private GitHub repo in a GoCD job. Wouldn't the repo url, version, and dir need to be passed into this task-creating method?

Comment (macdiesel) [inline edxpipelines/patterns/tasks.py]: This is true, you could only use one repository per pipeline if you needed this task. We could code the values directly in to the task if we think that would be better.

Comment (feanil) [inline edxpipelines/patterns/tasks.py]: Can we generalize this to support being able to run it multi times for multiple repos?

Comment (doctoryes) [inline edxpipelines/patterns/tasks.py]: I think Dillon's pipeline needs access to two different private repos already (edx-mtkg and acquia?) The way I had this coded last Friday was like this:
```
def _fetch_secure_repo(job, secure_dir, secure_repo_envvar, secure_version_envvar, secure_repo_name, runif="passed"):
    """
    Setup a secure repo for use in providing secrets.

    Args:
        job (gomatic.job.Job): the gomatic job to which the task will be added
        secure_dir (str): name of dir containing the edx-ops/configuration-secure repo
        secure_repo_envvar (str): HTTPS-based link to secure repo on GitHub
        secure_version_envvar (str): GitHub ref identifying version of secure repo to use
        secure_repo_name (str): name of secure repo, e.g. "configuration-secure"
        runif (str): one of ['passed', 'failed', 'any'] Default: passed

    Returns:
        The newly created task (gomatic.gocd.tasks.ExecTask)

    """
    return job.add_task(
        ExecTask(
            [
                '/bin/bash',
                '-c',
                'touch github_key.pem && '
                'chmod 600 github_key.pem && '
                'python tubular/scripts/format_rsa_key.py --key "$PRIVATE_GITHUB_KEY" --output-file github_key.pem && '
                "GIT_SSH_COMMAND='/usr/bin/ssh -o StrictHostKeyChecking=no -i github_key.pem' "
                '/usr/bin/git clone ${secure_repo_envvar} {secure_dir} && '
                'cd {secure_dir} && '
                '/usr/bin/git checkout ${secure_version_envvar} && '
                'mkdir ../target/ && '
                '/usr/bin/git rev-parse HEAD > ../target/{secure_repo_name}_sha'.format(
                    secure_dir=secure_dir,
                    secure_repo_envvar=secure_repo_envvar,
                    secure_version_envvar=secure_version_envvar,
                    secure_repo_name=secure_repo_name
                )
            ]
        )
    )
```
and then these methods calling it:
```
def fetch_secure_configuration(job, secure_dir, runif="passed"):
    """
    Setup the configuration-secure repo for use in providing secrets.

    Args:
        job (gomatic.job.Job): the gomatic job to which the playbook run task will be added
        secure_dir (str): name of dir containing the edx-ops/configuration-secure repo
        runif (str): one of ['passed', 'failed', 'any'] Default: passed

    Returns:
        The newly created task (gomatic.gocd.tasks.ExecTask)

    """
    return _fetch_secure_repo(
        job, secure_dir,
        "CONFIGURATION_SECURE_REPO",
        "CONFIGURATION_SECURE_VERSION",
        "configuration-secure"
    )


def fetch_gomatic_secure(job, secure_dir, runif="passed"):
    """
    Setup the gomatic-secure repo for use in providing secrets.

    Args:
        job (gomatic.job.Job): the gomatic job to which the task will be added
        secure_dir (str): name of dir containing the edx-ops/gomatic-secure repo
        runif (str): one of ['passed', 'failed', 'any'] Default: passed

    Returns:
        The newly created task (gomatic.gocd.tasks.ExecTask)

    """
    return _fetch_secure_repo(
        job, secure_dir,
        "GOMATIC_SECURE_REPO",
        "GOMATIC_SECURE_VERSION",
        "gomatic-secure"
    )
```

Comment (macdiesel) [inline edxpipelines/patterns/tasks.py]: My understanding is that the aquia would be a remote that would added to his git private repo.

@Dillon-Dumesnil?

Comment (Dillon-Dumesnil) [inline edxpipelines/patterns/tasks.py]: The acquia is a remote that needs to be added and then pushed to using a tag created from edx-mktg.

Comment (feanil) [inline edxpipelines/patterns/tasks.py]: I think we should add the generic function that @doctoryes wrote and then have this function call that.

**Comment by macdiesel** [comment]
@edx/pipeline-team Please take another pass. I cherry-picked john's implementation.

@Dillon-Dumesnil you will need to rebase from master once these changes are merged.

File: edxpipelines/patterns/tasks.py [hunk 2]
```diff
@@ -320,6 +326,48 @@ def generate_target_directory(job, directory_name="target", runif="passed"):
...
+        runif (str): one of ['passed', 'failed', 'any'] Default: passed
+
+    Returns:
+        The newly created task (gomatic.gocd.tasks.ExecTask)
+
+    """
+    return _fetch_secure_repo(
+        job, secure_dir,
+        "CONFIGURATION_SECURE_REPO",
+        "CONFIGURATION_SECURE_VERSION",
```
Comment (doctoryes) [inline edxpipelines/patterns/tasks.py]: Nit: Is it worth mentioning in the docstring that the caller needs to ensure `CONFIGURATION_SECURE_REPO` and `CONFIGURATION_SECURE_VERSION` are both defined as environment variables?

File: edxpipelines/patterns/tasks.py [hunk 3]
```diff
@@ -320,6 +326,48 @@ def generate_target_directory(job, directory_name="target", runif="passed"):
...
+        runif (str): one of ['passed', 'failed', 'any'] Default: passed
+
+    Returns:
+        The newly created task (gomatic.gocd.tasks.ExecTask)
+
+    """
+    return _fetch_secure_repo(
+        job, secure_dir,
+        "GOMATIC_SECURE_REPO",
+        "GOMATIC_SECURE_VERSION",
```
Comment (doctoryes) [inline edxpipelines/patterns/tasks.py]: Ditto nit.

File: edxpipelines/patterns/tasks.py (hunk 2 above)
Comment (macdiesel) [inline edxpipelines/patterns/tasks.py]: done

File: edxpipelines/patterns/tasks.py (hunk 3 above)
Comment (macdiesel) [inline edxpipelines/patterns/tasks.py]: done

**Comment by doctoryes** [comment]
👍 🐑 🇮🇹

**Comment by macdiesel** [comment]
baaaaa

**Pull request merged**